### PR TITLE
45 stake

### LINKED ISSUE #45: implement a `stake` function

### PR DESCRIPTION
**Pull request by jellegerbrandy**
This fixes #45

### CODE REVIEW

File: src/stake.ts [hunk 1]
```diff
@@ -52,22 +58,20 @@ export class Stake implements IStake {
...
-
   constructor(
       public id: string,
       public staker: string,
-      public createdAt: Date,
+      public createdAt: Date | undefined,
       public outcome: ProposalOutcome,
       public amount: number,
-      public proposalId: string,
-      public context: Arc
```
Comment (ben-kaufman) [inline src/stake.ts]: Why did you delete that (and what are the other changes here)?

**Comment by tsuberim** [comment]
Maybe I misread the code but, why do you send the transaction, and only validate it after it fails?  you should validate everything before sending the tx:
1. validate arguments client side.
2. compute gas limit.
3. use `.call` with real arguments and gas limit.
4. only then `.send` the tx after you are sure it will succeed.

Also this pattern is generic and should be extracted to a common function with only the first step being different for different operations.

**Comment by jellegerbrandy** [comment]
@tsuberim , you did not misread. 

To explain my reasoning: I am doing a post-portem on a failed transaction to generate some useful error messages (instead of the generic "revert" message from ethereum). These checks are relatively expensive (because they involve calls to the ethereum node), there could potentially be many of them, and so i'd like to avoid doing them _before_ sending the transaction if at all possible.

 The pattern I'm implementing here is:  do _some_ (minmal, cheap) argumetn validation before (basically whever you can do client side); and do a (more expesnive, in terms of performance) postmortem if the transaction fails. 

Notice that I'm not talking about gas costs here.  But, as you point out,  the "cost" of validation is not just   the cost of computation (i.e. wait for a request to the ethereum node), there is also the chance of wasting the gas in a failed transaction.

Now, I think that if alchemy is well written, it shoudl _never_ happen (or almost) that the app sends failing transactions to the chain. If such an error occur, it is  a bug in alchemy. The error handling is for developers debugging their code, not end users. So in that logic, there is never any wasted gas, and so error handlign should be done post-hoc to save on performance.

So that was my reasoning. Does it make sense? 

(small question: what do yo mean by "compute gas limit"?)






 My reasoning was more or less like this. There is "cheap validation" (cheap in terms of performance, not gas costs) such as checking of all args are there, and there is "expensive validation" (again in terms of perforamnce) which requires calls to the blockchain.

Now, I'd like to avoid doing too much expesnive stuff by default before sending. Especially, for complex transactions, this would mean checking balances

**Comment by tsuberim** [comment]
I see your point, but I think avoiding actual funds loss is more important than the performance overhead of checking the conditions. Of course you can always not do any validation and blame the user for sending bad arguments, but in our case that could mean losing some real money.

Re "compute gas limit": web3 can compute how much gas a tx is going to consume beforehand, you can use this and automatically set the gas limit for the tx.

**Comment by jellegerbrandy** [comment]
I am not sure you see my point :-) 

> I think avoiding actual funds loss is more important than the performance overhead of checking the conditions

Stated like that, it seems undisputable: "I thnk avoiding an actual loss of 10 cents on a transaction is more important than spending 0,001 cents in resources to check the conditions".

But what I am saying is that "I think avoiding a 0,001% chance of loosing 10 cents is less important then spending  times 0,01  cents in resources to check preconditions"  - multiplied by X requests per day.

Now the user osts are hard to quantify (there is some miniscule costs in  resources (i.e. sending the requests and calculating the answers) and some hard-to-quantify costs in user time spend waiting longer for trnasactions to go through.

We decided when staring out this client library project that we'd not try to develop a generic solution (at least not in the first place), but a solution geared to the needs of alchemy. So, "blame the user" in this case means "blame the developers of alchemy". Alchemy should not send invalid requests, period. If alchemy tries to send an invalid request, we shoudl fix the bug, not wast the user's time with error checking (and showing him the error but not sending the transaction)

So I'm still for the post-portem checking,

We could also consider an option like this but I'd hate to introduce this kind of complexity at this point:

```
proposal.vote(YES) // creates the subscribable, but does nothing
propoval.vote(YES).checkPreconditions() // check the preConditions
propoval.vote(YES).subscribe(....) // send the transction, on failure checks preconditions
```
And perhaps even an configuration option in `arc` to "always check preconditions".

**Comment by jellegerbrandy** [comment]
Thinking about tis a bit more (clearly overthinking this :-)), @tsuberim, I could implement your  generic pre-check anyway (by `call`ing the transactino bf sending it), and then do the post-mortem only if it fails (i.e. in exactly the same place as that I do it now). It is just a single call to the blockchain, then, so I suppose it's worth it.

**Comment by jellegerbrandy** [comment]
i created this issue for the generic validation stuff, as it will take me some time to figure out how to weave the async calls into the observable pattern: https://github.com/daostack/client/issues/53

Otherwise, if there are no objections, can someone approve this PR so we can continue building on it?

**Comment by tsuberim** [comment]
@jellegerbrandy 
I changed my mind about this, but not because of your reasons :). A few comments:
1. The "rational" user would indeed take into account the average cost of the operation, but people don't actually think that way. People have a bias to prefer safe incentive and risky disincentives.
2. I now think having a performent app is more important than the funds because:
   a. The loss of funds is relatively small and constant. 
   b. There may be circumstances where the user wants to get a tx mined ASAP and having it delayed may cause a far greater loss of funds (or other troubles).
   c. A performent app is important for getting the system adopted at all.
3. 
     > We decided when staring out this client library project that we'd not try to develop a generic solution (at least not in the first place), but a solution geared to the needs of alchemy.
     
     I agree, but I think you may be abusing the statement a bit. Taking your approach to the limit, you should just develop alchemy without bothering about a client lib. My view is: develop only features that alchemy will use, but develop them such that other people can also use them.

**Comment by jellegerbrandy** [comment]
Ok @tsuberim , I am now (after your arguments) of the opinion that we should do the generic `call() works?` check anyway - it is relatively cheap and fast (should not delay the sending of the transaction significantly). In any case, we can develop that later, and if needed discuss this further in #53,  and in the meanwhile **approve this request** :-)
